03-wordemb-pytorch/wordemb-cbow.py

```diff
@@ -0,0 +1,122 @@
+from collections import defaultdict
+import math
+import time
+import random
+import torch
+
+
+class CBoW(torch.nn.Module):
+    def __init__(self, nwords, emb_size):
+        super(CBoW, self).__init__()
+
+        """ layers """
+        self.embedding = torch.nn.Embedding(nwords, emb_size)
+        """ projection layer for taking softmax over vocabulary words"""
+        self.projection = torch.nn.Linear(emb_size, nwords)
+
+    def forward(self, words):
+        emb = self.embedding(words)
+        emb_sum = torch.sum(emb, dim=0)  # size(emb_sum) = emb_size
+        emb_sum = emb_sum.view(1, -1)  # size(emb_sum) = 1 x emb_size
+        out = self.projection(emb_sum)  # size(out) = 1 x nwords
+        return out
+
+
+N = 2  # length of window on each side (so N=2 gives a total window size of 5, as in t-2 t-1 t t+1 t+2)
+EMB_SIZE = 128  # The size of the embedding
+
+embeddings_location = "embeddings.txt"  # the file to write the word embeddings to
+labels_location = "labels.txt"  # the file to write the labels to
+
+# We reuse the data reading from the language modeling class
+w2i = defaultdict(lambda: len(w2i))
+S = w2i["<s>"]
+UNK = w2i["<unk>"]
+
+
+def read_dataset(filename):
+    with open(filename, "r") as f:
+        for line in f:
+            yield [w2i[x] for x in line.strip().split(" ")]
+
+
+# Read in the data
+train = list(read_dataset("../data/ptb/train.txt"))
+w2i = defaultdict(lambda: UNK, w2i)
+dev = list(read_dataset("../data/ptb/valid.txt"))
+i2w = {v: k for k, v in w2i.items()}
+nwords = len(w2i)
+
+with open(labels_location, 'w') as labels_file:
+    for i in range(nwords):
+        labels_file.write(i2w[i] + '\n')
+
+# initialize the model
+model = CBoW(nwords, EMB_SIZE)
+criterion = torch.nn.CrossEntropyLoss()
+optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
+
+type = torch.LongTensor
+use_cuda = torch.cuda.is_available()
+
+if use_cuda:
+    type = torch.cuda.LongTensor
+    model.cuda()
+
+
+# Calculate the loss value for the entire sentence
+def calc_sent_loss(sent, inference=False):
+
+    # add padding to the sentence equal to the size of the window
+    # as we need to predict the eos as well, the future window at that point is N past it
+    padded_sent = [S] * N + sent + [S] * N
+
+    # Step through the sentence
+    total_loss = 0
+    for i in range(N, len(sent) + N):
+        # c is the context vector
+        c = torch.tensor(padded_sent[i - N:i] + padded_sent[i + 1:i + N + 1]).type(type)
+        t = torch.tensor([padded_sent[i]]).type(type) # This is the target vector
+        log_prob = model(c)
+        loss = criterion(log_prob, t)   # loss for predicting target from context vector
+        if not inference:
```

Comment:
in the Dynet example, the backprop happens after summing all the losses for a given sentence rather than after every word seen, I think it would be best to make it consistent with that.

Reply:
Also, I think the choice of `inference` is confusing. Technically, you are doing inference during training and dev both. Maybe just change `inference` flag with the opposite `need_to_train` to avoid the confusion.

Reply:
@danishpruthi : Thanks for pointing this out. I have moved the backward operation in the main loop, thus the flag is no longer used.